readme.txt
=== ElasticPress ===
Contributors: tlovett1, vhauri, tott, oscarssanchez, cmmarslender, 10up
Author URI: http://10up.com
Plugin URI: https://github.com/10up/ElasticPress
Tags: performance, slow, search, elasticsearch, fuzzy, facet, aggregation, searching, autosuggest, suggest, elastic, advanced search, woocommerce, related posts, woocommerce
Requires at least: 3.7.1
Tested up to: 5.5
Stable tag: trunk
License: GPLv2 or later
License URI: http://www.gnu.org/licenses/gpl-2.0.html

A fast and flexible search and query engine for WordPress.

== Description ==
ElasticPress, a fast and flexible search and query engine for WordPress, enables WordPress to find or “query” relevant content extremely fast through a variety of highly customizable features. WordPress out-of-the-box struggles to analyze content relevancy and can be very slow. ElasticPress supercharges your WordPress website making for happier users and administrators. The plugin even contains features for popular plugins.

Here is a list of the amazing ElasticPress features included in the plugin:

__Search__: Instantly find the content you’re looking for. The first time.

__WooCommerce__: “I want a cotton, woman’s t-shirt, for under $15 that’s in stock.” Faceted product browsing strains servers and increases load times. Your buyers can find the perfect product quickly, and buy it quickly.

__Related Posts__: ElasticPress understands data in real time, so it can instantly deliver engaging and precise related content with no impact on site performance.

__Protected Content__: Optionally index all of your content, including private and unpublished content, to speed up searches and queries in places like the administrative dashboard.

__Documents__: Indexes text inside of popular file types, and adds those files types to search results.

__Autosuggest__: Suggest relevant content as text is entered into the search field.

__Facets__: Add controls to your website to filter content by one or more taxonomies.

__Users__: Improve user search relevancy and query performance.

Please refer to [Github](https://github.com/10up/ElasticPress) for detailed usage instructions and documentation.

== Installation ==
1. First, you will need to properly [install and configure](https://www.elastic.co/guide/en/elasticsearch/reference/current/setup.html) Elasticsearch.
2. Activate the plugin in WordPress.
3. In the ElasticPress settings page, input your Elasticsearch host.
4. Sync your content by clicking the sync icon.
5. Enjoy!

== Changelog ==

= 3.4.4 =

Bug Fixes:
* Add required (since WP 5.5) permission callback to register_rest_route() methods.

= 3.4.3 =

Enhancements:
* Remove jQuery from front end JavaScript dependencies.

Bug Fixes:
* Fix accessibility bug on autosuggest.
* Fix broken facet search.

= 3.4.2 =

Bug fixes:
* uninstall.php: Change the EP_FILE const to its value. Props [felipeelia](https://github.com/felipeelia).
* Fix list features WP CLI command. Props [felipeelia](https://github.com/felipeelia).
* Add `rel="nofollow"` to facet links. Props [mlaroy](https://github.com/mlaroy).
* Facets widget: Move <div> outside ob_start(). Props [kallehauge](https://github.com/kallehauge).
* Load facet scripts and styles only when they are really necessary. Props [goaround](https://github.com/goaround).
* Index attachments with Protected Content and query for them in media search. Props [oscarsanchez](https://github.com/oscarsanchez).
* Fixed `Deprecated field [include] used, expected [includes] instead.`. Props [dinhtungdu](https://github.com/dinhtungdu).

Enhancements:
* Add filter for enabling sticky posts.  Props [shadyvb](https://github.com/shadyvb).
* Add sync kill filter. Props [barryceelen](https://github.com/barryceelen).
* Add timeout filters for bulk_index and index_document. Props [@oscarsanchez](https://github.com/oscarsanchez).

= 3.4.1 =

* Make weighting dashboard flex containers to prevent the slider from changing size. Props [@mlaroy](https://github.com/mlaroy).
* Fix issue where weightings wouldn't save properly for certain post types. Props [mustafauysal](https://github.com/mustafauysal).
* Fix bug where terms wouldn't finish syncing in certain scenarios.
* Properly order WooCommerce products using double to account for decimals. Props [@oscarsanchez](https://github.com/oscarsanchez).
* Show current indices in index health dashboard. Props [moraleida](https://github.com/moraleida).

= 3.4 =

* Addition of Terms Indexable and Feature. ElasticPress can now integrate with `WP_Term_Query`. Props [dkotter](https://github.com/dkotter).
* Fixes for `WP_User_Query` 'fields' parameter. Props [petenelson](https://github.com/petenelson).
* Support all taxonomies in root of `WP_Query`
* Readd `ep_retrieve_aggregations` hook for backwards compatibility
* Move indexable posts class registration into a function that runs in the `plugins_loaded` action. Props [petenelson](https://github.com/petenelson).
* Fix author name in weighting and use post_author.display_name for weighted author field search. Props [petenelson](https://github.com/petenelson).
* Add `ep_prepared_*_meta` filters
* Refactor CLI indexing code for simplicity.
* Limit indexed Protected Content post types removing `revision`, `oembed_cache`, `custom_css`, `user_request`, `customize_changeset`, and `wp_block`.
* Cast taxonomy slug to array in case it's already an array in `WP_Query`.
* Remove unnecessary usage of `--network-wide` CLI paramter.
* Add name, nickname, and display name to fields used for user search.
* Add `clear-transient` WP CLI command.
* Don't make product categories facetable when WooCommerce feature is not active. Props [mustafauysal](https://github.com/mustafauysal).

= 3.3 =

* Officially support Elasticsearch 7.5
* Add optional Google Analytics Autosuggest tracking Event
* Fix single node warning before sync has occurred.
* When `ep_integrate` is set to false, do not apply faceting.
* Fix search ordering error when there are no pointers.
* Add filter `ep_get_hits_from_query` to modify retrieved Elasticsearch hits.
* Make sure `post_type` array does not include keys. Fixes a bbPress issue.
* Pass query object to EP response so we can check for main query. This fixes a faceting bug.
* Add EP-Search-Term header to autosuggest requests to EP.io
* Clean up indexing transient on sigkill

= 3.2.6 =
This is a bugfix release

* Under some edge conditions content for autosuggest can be large - don't cache it

= 3.2.5 =
This is a bug fix version.

* Fix WP <5.0 fatal error on register_block_type.

= 3.2.4 =
This is a bug fix version.

* Fix Gutenberg block initialization
* Fix Autosuggest: remove filter with proper priority in query generation. Props [Maxdw](https://github.com/Maxdw).
* Fix Autosuggest: returning WP_Error for non object cache autosuggest queries causes issue. Fallback to transient

= 3.2.3 =
This is a bug fix version.

* Ensure query building for Autosuggest does not fallback to WPDB.

= 3.2.2 =
This is a bug fix version with some feature additions.

* Fix PHPCS errors. Props [mmcachran](https://github.com/mmcachran)
* Fix ensuring stats are built prior to requesting information
* Fix related post block enqueue block assets on the frontend
* Fix custom order results change webpack config for externals:lodash
* Fix don't overwrite search fields
* Autosuggest queries generated though PHP instead of JavaScript
* Add WP Acceptance tests
* Add new WP-CLI commands: get_indexes and get_cluster_indexes

= 3.2.1 =
This is a bug fix version.

* Fix Gutenberg breaking issue with Related Posts and image blocks. Props [adamsilverstein](https://github.com/adamsilverstein)

= 3.2 =
ElasticPress 3.2 is a feature release. We've added quite a few useful features including an index health page, the ability to control which sites are indexed in a network activated multisite setup, a related posts Gutenberg block, and more.

* Improve block asset enqueueing: hook on `enqueue_block_editor_assets`. Props [adamsilverstein](https://github.com/adamsilverstein).
* Handle empty search weighting fields bug.
* Integrate WooCommerce default filter by price widget with ES range query.
* Improve messaging for custom result post type.
* Index health page.
* Add tag_in and tag__and support.
* Related posts Gutenberg block.
* Facet widget ordering option. Props [psorensen](https://github.com/psorensen).
* Control Index-ability of individual sites in multisite.
* Integrate WooCommerce default filter by price widget with ES range query.

See https://github.com/10up/ElasticPress/pulls?utf8=%E2%9C%93&q=is%3Apr+milestone%3A3.2.0+is%3Aclosed+

= 3.1.4 =
https://github.com/10up/ElasticPress/pulls?q=is%3Apr+milestone%3A3.1.4+is%3Aclosed

= 3.1.3 =
This is a bug fix release.

* Check wpcli transient before integrating with queries
* Fix version comparison bug when comparing Elasticsearch versions
* Use proper taxonomy name for WooCommerce attributes.
* Increase Elasticsearch minimum supported version to 5.0
* Fix product attribute archives

= 3.1.2 =
This is a bug fix release with some filter additions.

- Add ep_es_query_results filter.
- Add option to sync prior to shutdown.
- Readme update around WPCLI post syncing. Props [@mmcachran](https://github.com/mmcachran)
- Ignore sticky posts in `find_related`. Props [@columbian-chris](https://github.com/columbian-chris)
- Weighting dashboard fixes around saving. [@oscarsanchez](https://github.com/oscarsanchez)
- Weighting UI improvements. Props [@mlaroy](https://github.com/mlaroy)

= 3.1.1 =
- Ensure taxonomies that are shared among multiple post types show up on the weighting screen

= 3.1.0 =

- Support for nested tax queries. Props [@dkotter](https://github.com/dkotter)
- `ep_bulk_index_action_args` filter. Props [@fabianmarz](https://github.com/fabianmarz)
- Add filters to control MLT related posts params.
- `ep_allow_post_content_filtered_index` filter to bypass filtered post content on indexing.
- Weighting dashboard to control weights of specific fields on a per post type basis
- Search ordering feature. Enables custom results for specific search queries.
- Refactor admin notice, admin screen "resolver", and install path logic
- WordPress.org profile
- New EP settings interface. Props [@dkoo](https://github.com/dkoo)
- Delete pagination from facet URL.
- allows WooCommerce product attributes to be facetable in 3.0
- Autosuggest queries now match the search queries performed by WordPress, including weighting and any custom results
- Fix data escaping in WP 4.8.x
- Support order by "type"/"post_type" in EP queries
- Properly redirect after network sync
- User mapping for pre 5.0 Props [@mustafauysal](https://github.com/mustafauysal)
- Avoid multiple reflows in autosuggest. Props [@fabianmarz](https://github.com/fabianmarz)
- 400 error when popularity is default sorting.
- Fixed Facet widget not rendering WC product attribute options. Props [@fabianmarz](https://github.com/fabianmarz)
- Delete wpcli sync option/transient when an error occurs
- Create index/network alias when adding a new site on a network activated installation. Props [@elliott-stocks](https://github.com/elliott-stocks)
- Fix WooCommerce order search when WooCommerce module activated but protected content turned off.

= 3.0.3 =

* Pass $post_id twice in ep_post_sync_kill for backwards compatibility. Props [aaemnnosttv](https://github.com/aaemnnosttv)
* Add `ep_search_request_path` filter for backwards compant.
* Add `ep_query_request_path` filter for modifying the query path.
* Fix missing action name in post query integration.
* Properly add date filter to WP_Query.

= 3.0.2 =

3.0.2 is a minor bug release version. Here is a list of fixes:

* Fix date query errors
* Readd ep_retrieve_the_{type} filter. Props [gassan](https://github.com/gassan)
* Fix empty autosuggest selector notice

= 3.0.1 =

3.0.1 is a minor bug release version. Here is a list of fixes:

* `wp elasticpress stats` and `wp elasticpress status` commands fatal error fixed.
* Add autosuggest selector field default to fix notice.
* Re-add `ep_find_related` as deprecated function.
* Changed max int to use core predefined constant. Props [@fabianmarz](https://github.com/fabianmarz)
* Properly support legacy feature registration callbacks per #1329.
* Properly disable settings as needed on dashboard.
* Don't force document search on REST requests.

= 3.0 (Requires re-index) =

3.0 is a refactor of ElasticPress for modern coding standards (PHP 5.4 required) as well as the introduction to indexables. Indexables abstracts out content types so data types other than post can be indexed and searched. 3.0 includes user indexing and search (integration with WP_User_Query). User features require at least WordPress version 5.1.

The refactor changes a lot of ElasticPress internals. The biggest change is the feature registration API has completely changed. Now, new features should extend the `ElasticPress\Feature` class rather than calling `ep_register_feature`. Older features should be backwards compatible.

Other Features:
* Elasticsearch language setting in admin

Here are a list of filters/actions removed or changed:

### Actions Removed:

* `ep_feature_setup`

### Filters changed:

* `ep_post_sync_kill` - Removed `$post_args` argument.

### Other changes:

* `posts-per-page` changed to `per-page` for WP-CLI index command.

= 2.8.2 =

* Enhancement: WooCommerce product attributes as facets.
* Enhancement: Performance Boost for document indexing.
* Bugfix for issue on WP REST API searches.
* Bugfix for case-sensitivity issue with facet search.

= 2.8.1 =

* Bugfix for homepage out of chronological order.
* Bugfix for missing meta key. (Props [turtlepod](https://github.com/turtlepod))
* Bugfix for bulk indexing default value on settings page.

= 2.8.0 =

ElasticPress 2.8 provides some new enhancements and bug fixes.

* Sticky posts support.
* Meta LIKE query adjustment.
* Autosuggest bugfix.
* Autosuggest to abide by plugin settings.
* WooCommerce searches with custom fields.
* Adjustment to `wp elasticpress status`
* Add Elasticsearch version in settings. (Props [turtlepod](https://github.com/turtlepod))
* Allow user to set number of posts during bulk indexing cycle.
* Facet query string customization (Props [ray-lee](https://github.com/ray-lee))
* Removal of logic that determines if blog is public / indexable. (Resolves sync issue.)
* Fix for auto activating sync notices. (Props [petenelson](https://github.com/petenelson))
* Removal of date weighting for protected content admin queries.
* Protected content: filtering of filtered post types.
* Implemented --post-ids CLI option to index only specific posts. (Props [dotancohen](https://github.com/dotancohen))

= 2.7.0 (Requires re-index) =

ElasticPress 2.7 provides some new enhancements and bug fixes.

* Prevent indexing when blog is deleted or not public.
* Do not apply absint to comment_status.
* ElasticPress.io credentials bugfix.
* Related posts bugfix.
* Random WooCommerce ordering allowed.
* Query only post IDs when indexing. (Props [elliott-stocks](https://github.com/elliott-stocks))
* Better error notices. (Props [petenelson](https://github.com/petenelson))

= 2.6.1 =

* Resolves issue of missing file for wp-cli.

= 2.6.0 =

ElasticPress 2.6 provides some new enhancements and bug fixes.

* Ability to set autosuggest endpoint by a constant (EP_AUTOSUGGEST_ENDPOINT).
* Enable WooCommerce products to be included in autosuggest results.
* Support for tax_query operators EXISTS and NOT EXISTS.
* Addition of new filter to change default orderby/sort (ep_set_default_sort).
* Do not search for author_name when searching products in WooCommerce.

= 2.5.2 (Requires re-index) =

This is a small bug fix release.

* Removed unnecessary facet JavaScript
* Fix facet aggregations warning

= 2.5.1 (Requires re-index) =

This if a bug fix release. This version requires a re-index as we change the way data is being sent to Elasticsearch.

It's also worth noting for ElasticPress version 2.5+, the Facets feature, which is on by default, will run post type archive and search page main queries through Elasticsearch. If Elasticsearch is out of sync with your content (possible in rare edge cases), this could result in incorrect content being shown. Turning off Facets would fix the problem.

### Bug Fixes

* Don't pre-strip HTML before sending it to Elasticsearch.
* Support PHP 5.2 backwards compat.
* Don't show faceting widget if post type doesn't support taxonomy.

= 2.5 =

ElasticPress 2.5 includes a new Facets feature that makes it easy to add high performance content filtering controls to a website.

A new Facets widget enables site administrators to add taxonomy facets to a sidebar (or any widgetized area). When viewing a content list on the front end of the website, the widget will display the name of the taxonomy – e.g. “Categories” – and a checklist with all of its terms. Visitors can narrow down content by selecting terms they are interested in. The Facets feature can be globally configured to narrow results to content that is tagged with any or all of the selected terms. The widget’s front end output contains carefully named CSS classes, so that designers and developers can apply unique styling.

Version 2.5 also includes a number of smaller enhancements and fixes, including official support for Elasticsearch 6.2, and increased functional parity with the WP_Query API.

Here is a detailed list of what's been included in the release:

### Enhancements
* Facets feature
* `--post-ids` CLI option to index only specific posts. Props [dotancohen](https://github.com/dotancohen).
* Filter for hiding host setting in dashboard. Props [tomdxw](https://github.com/tomdxw).
* Support `WP_Query` meta query `not between` comparator.

### Bugs
* Disallow duplicated Elasticsearch requests on WooCommerce orders page. Props [lukaspawlik](https://github.com/lukaspawlik)
* Fix taxonomy sync object warning. Props [eugene-manuilov](https://github.com/eugene-manuilov)
* `true` in `is_empty_query` terminates ep_query process when it shouldn't. Props [yaronuliel](https://github.com/yaronuliel)

= 2.4.2 =
Version 2.4.2 is a bug fix version.

* Fix related posts not showing up bug.

= 2.4.1 =

Version 2.4.1 is a bug fix and maintenance release. Here are a listed of issues that have been resolved:

* Support Elasticsearch 6.1 and properly send Content-Type header with application/json. Props [eugene-manuilov](https://github.com/eugene-manuilov).
* Fix autosuggest event target issue bug. Props [eugene-manuilov](https://github.com/eugene-manuilov).
* Fix widget init bug. Props [eugene-manuilov](https://github.com/eugene-manuilov).
* Fix taxonomy sync parameter warning. Props [eugene-manuilov](https://github.com/eugene-manuilov).
* Increase maximum Elasticsearch compatibility to 6.1

= 2.4 =

Version 2.4 introduces the Autosuggest feature. When enabled, input fields of type "search" or with the CSS class "search-field" or "ep-autosuggest" will be enhanced with autosuggest functionality. As text is entered into the search field, suggested content will appear below it, based on top search results for the text. Suggestions link directly to the content.

We also added hooks and filters to ElasticPress that make query logging possible. The [Debug Bar ElasticPress](https://github.com/10up/debug-bar-elasticpress) plugin now adds a Query Log screen to the ElasticPress admin menu. The Query Log is an extremely powerful tool for diagnosing search and indexing issues.

Here is a comphrensive list of changes:

### Enhancements
* Autosuggest feature
* Hooks for query log functionality in [Debug Bar ElasticPress](https://github.com/10up/debug-bar-elasticpress)
* Support `WP_Query` `fields` parameter. Props [kallehauge](https://github.com/kallehauge).
* Add setting for enabling/disabling date weighting in search. Props [lukaspawlik](https://github.com/kallehauge).
* Remove extra post meta storage key from Elasticsearch
* Add shipping class as indexed WooCommerce taxonomy. Props [kallehauge](https://github.com/kallehauge).
* Allow WooCommerce orders to be searched by items. Props [kallehauge](https://github.com/kallehauge).
* Support Elasticsearch 5.6
* Add filter to granularly control admin notices. Props [mattonomics](https://github.com/mattonomics).
* Support ES 5.5+ strict content type checking. Props [sc0ttclark](https://github.com/sc0ttclark)

### Bug Fixes
* Fix `author_name` search field. Props [ivankristianto](https://github.com/ivankristianto).
* Fix unavailable taxonomy issue in WooCommerce. Props [ivankristianto](https://github.com/ivankristianto).
* Index all publicly queryable taxonomies. Props [allan23](https://github.com/allan23).
* Resolve case insensitive sorting issues. Props [allan23](https://github.com/allan23).
* Add escaping per VIP standards. Props [jasonbahl](https://github.com/jasonbahl).
* Fix WooCommerce post type warnings.

= 2.3.1, 2.3.2 =

Version 2.3.1-2.3.2 is a bug fix release. Here are a listed of issues that have been resolved:


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file, per [the Keep a Changelog standard](http://keepachangelog.com/).

## [3.4.4] - 2020-08-25

Bug Fixes:

- Add required (since WP 5.5) permission callback to register_rest_route() methods.

## [3.4.3] - 2020-07-21

Enhancements:

- Remove jQuery from front end JavaScript dependencies.

Bug Fixes:

- Fix accessibility bug on autosuggest.
- Fix broken facet search.

## [3.4.2] - 2020-06-17

Bug fixes:

- uninstall.php: Change the EP_FILE const to its value. Props [felipeelia](https://github.com/felipeelia).
- Fix list features WP CLI command. Props [felipeelia](https://github.com/felipeelia).
- Add `rel="nofollow"` to facet links. Props [mlaroy](https://github.com/mlaroy).
- Facets widget: Move <div> outside ob_start(). Props [kallehauge](https://github.com/kallehauge).
- Load facet scripts and styles only when they are really necessary. Props [goaround](https://github.com/goaround).
- Index attachments with Protected Content and query for them in media search. Props [oscarsanchez](https://github.com/oscarsanchez).
- Fixed `Deprecated field [include] used, expected [includes] instead.`. Props [dinhtungdu](https://github.com/dinhtungdu).

Enhancements:

- Add filter for enabling sticky posts. Props [shadyvb](https://github.com/shadyvb).
- Add sync kill filter. Props [barryceelen](https://github.com/barryceelen).
- Add timeout filters for bulk_index and index_document. Props [@oscarsanchez](https://github.com/oscarsanchez).

## [3.4.1] - 2020-3-31

- Make weighting dashboard flex containers to prevent the slider from changing size. Props [@mlaroy](https://github.com/mlaroy).
- Fix issue where weightings wouldn't save properly for certain post types. Props [mustafauysal](https://github.com/mustafauysal).
- Fix bug where terms wouldn't finish syncing in certain scenarios.
- Properly order WooCommerce products using double to account for decimals. Props [@oscarsanchez](https://github.com/oscarsanchez).
- Show current indices in index health dashboard. Props [moraleida](https://github.com/moraleida).

## [3.4]

- Addition of Terms Indexable and Feature. ElasticPress can now integrate with `WP_Term_Query`. Props [dkotter](https://github.com/dkotter).
- Fixes for `WP_User_Query` 'fields' parameter. Props [petenelson](https://github.com/petenelson).
- Support all taxonomies in root of `WP_Query`
- Readd `ep_retrieve_aggregations` hook for backwards compatibility
- Move indexable posts class registration into a function that runs in the `plugins_loaded` action. Props [petenelson](https://github.com/petenelson).
- Fix author name in weighting and use post_author.display_name for weighted author field search. Props [petenelson](https://github.com/petenelson).
- Add `ep_prepared_*_meta` filters
- Refactor CLI indexing code for simplicity.
- Limit indexed Protected Content post types removing `revision`, `oembed_cache`, `custom_css`, `user_request`, `customize_changeset`, and `wp_block`.
- Cast taxonomy slug to array in case it's already an array in `WP_Query`.
- Remove unnecessary usage of `--network-wide` CLI paramter.
- Add name, nickname, and display name to fields used for user search.
- Add `clear-transient` WP CLI command.
- Don't make product categories facetable when WooCommerce feature is not active. Props [mustafauysal](https://github.com/mustafauysal).

## [3.3] - 2018-12-18

- Officially support Elasticsearch 7.5
- Add optional Google Analytics Autosuggest tracking Event
- Fix single node warning before sync has occurred.
- When `ep_integrate` is set to false, do not apply faceting.
- Fix search ordering error when there are no pointers.
- Add filter `ep_get_hits_from_query` to modify retrieved Elasticsearch hits.
- Make sure `post_type` array does not include keys. Fixes a bbPress issue.
- Pass query object to EP response so we can check for main query. This fixes a faceting bug.
- Add EP-Search-Term header to autosuggest requests to EP.io
- Clean up indexing transient on sigkill

## [3.2.6] - 2019-11-27

- Fix Under some edge conditions content for autosuggest can be large - don't cache it

## [3.2.5] - 2019-11-20

- Fix WP <5.0 fatal error on register_block_type.

## [3.2.4] - 2019-11-19

- Fix Gutenberg block initialization
- Fix Autosuggest: remove filter with proper priority in query generation. Props [Maxdw](https://github.com/Maxdw).
- Fix Autosuggest: returning WP_Error for non object cache autosuggest queries causes issue. Fallback to transient

## [3.2.3] - 2019-11-13

- Ensure query building for Autosuggest does not fallback to WPDB.

## [3.2.2] - 2019-11-05

- Fix PHPCS errors. Props [mmcachran](https://github.com/mmcachran)
- Fix ensuring stats are built prior to requesting information
- Fix related post block enqueue block assets on the frontend
- Fix custom order results change webpack config for externals:lodash
- Fix don't overwrite search fields
- Autosuggest queries generated though PHP instead of JavaScript
- Add WP Acceptance tests
- Add new WP-CLI commands: get_indexes and get_cluster_indexes

## [3.2.1] - 2019-10-14

- Fix Gutenberg breaking issue with Related Posts and image blocks. Props [adamsilverstein](https://github.com/adamsilverstein)

## [3.2] - 2019-10-08

- Improve block asset enqueueing: hook on `enqueue_block_editor_assets`. Props [adamsilverstein](https://github.com/adamsilverstein).
- Handle empty search weighting fields bug.
- Integrate WooCommerce default filter by price widget with ES range query.
- Improve messaging for custom result post type.
- Index health page.
- Add tag_in and tag\_\_and support.
- Related posts Gutenberg block.
- Facet widget ordering option. Props [psorensen](https://github.com/psorensen).
- Control Index-ability of individual sites in multisite.
- Integrate WooCommerce default filter by price widget with ES range query.

See https://github.com/10up/ElasticPress/pulls?utf8=%E2%9C%93&q=is%3Apr+milestone%3A3.2.0+is%3Aclosed+


includes/classes/Feature/RelatedPosts/RelatedPosts.php
<?php

/**
					 * ElasticPress related posts feature
					 *
					 * @since  2.1
					 * @package elasticpress
					 */

namespace ElasticPress\Feature\RelatedPosts;

use ElasticPress\Feature as Feature;
use ElasticPress\Elasticsearch as Elasticsearch;
use ElasticPress\Post\Post as Post;
use \WP_Query as WP_Query;

/**
 * Related posts feature class
 */
class RelatedPosts extends Feature
{
	/**
	 * Initialize feature setting it's config
	 *
	 * @since  3.0
	 */
	public function __construct()
	{
		$this->slug = 'related_posts';

		$this->title = esc_html__('Related Posts', 'elasticpress');

		$this->requires_install_reindex = false;

		parent::__construct();
	}

	/**
	 * Format args for related posts
	 *
	 * @param  array $formatted_args Formatted ES args
	 * @param  array $args WP_Query args
	 * @return array
	 */
	public function formatted_args($formatted_args, $args)
	{
		if (!empty($args['more_like'])) {
			// lets compare ES version to see if new MLT structure applies
			$new_mlt = version_compare(Elasticsearch::factory()->get_elasticsearch_version(), 6.0, '>=');

			if ($new_mlt && is_array($args['more_like'])) {
				foreach ($args['more_like'] as $id) {
					$ids[] = array('_id' => $id);
				}
			} elseif ($new_mlt && !is_array($args['more_like'])) {
				$ids = array('_id' => $args['more_like']);
			} else {
				$ids = is_array($args['more_like']) ? $args['more_like'] : array($args['more_like']);
			}

			$mlt_key = ($new_mlt) ? 'like' : 'ids';

			$formatted_args['query'] = array(
				'more_like_this' => array(
					$mlt_key          => $ids,
					/**
					 * Filter fields used to determine related posts
					 *
					 * @hook ep_related_posts_fields
					 * @param  {array} $fields Related post fields
					 * @return  {array} New fields
					 */
					'fields'          => apply_filters(
						'ep_related_posts_fields',
						array(
							'post_title',
							'post_content',
							'terms.post_tag.name',
						)
					),
					/**
					 * Filter related posts minimum term frequency
					 *
					 * @hook ep_related_posts_min_term_freq
					 * @param  {int} $minimum Minimum term frequency
					 * @return  {array} New value
					 */
					'min_term_freq'   => apply_filters('ep_related_posts_min_term_freq', 1),
					/**
					 * Filter related posts maximum query terms
					 *
					 * @hook ep_related_posts_max_query_terms
					 * @param  {int} $maximum Maximum query terms
					 * @return  {array} New value
					 */
					'max_query_terms' => apply_filters('ep_related_posts_max_query_terms', 12),
					/**
					 * Filter related posts minimum document frequency
					 *
					 * @hook ep_related_posts_min_doc_freq
					 * @param  {int} $minimum Minimum document frequency
					 * @return  {array} New value
					 */
					'min_doc_freq'    => apply_filters('ep_related_posts_min_doc_freq', 1),
				),
			);
		}

		return $formatted_args;
	}

	/**
	 * Search Elasticsearch for related content
	 *
	 * @param  int $post_id Post ID
	 * @param  int $return Return code
	 * @since  2.1
	 * @return array|bool
	 */
	public function find_related($post_id, $return = 5)
	{
		$args = array(
			'more_like'           => $post_id,
			'posts_per_page'      => $return,
			'ep_integrate'        => true,
			'ignore_sticky_posts' => true,
		);

		/**
		 * Filter WP Query related post arguments
		 *
		 * @hook ep_find_related_args
		 * @param  {array} $args WP Query arguments
		 * @since  2.1
		 * @return  {array} New arguments
		 */
		$query = new WP_Query(apply_filters('ep_find_related_args', $args));

		if (!$query->have_posts()) {
			return false;
		}
		return $query->posts;
	}

	/**
	 * Setup all feature filters
	 *
	 * @since  2.1
	 */
	public function setup()
	{
		add_action('widgets_init', [$this, 'register_widget']);
		add_filter('ep_formatted_args', [$this, 'formatted_args'], 10, 2);
		add_action('init', [$this, 'register_block']);
		add_action('rest_api_init', [$this, 'setup_endpoint']);
	}

	/**
	 * Setup REST endpoints
	 *
	 * @since  3.2
	 */
	public function setup_endpoint()
	{
		register_rest_route(
			'wp/v2',
			'/posts/(?P<id>[0-9]+)/related',
			[
				'methods'  => 'GET',
				'callback' => [$this, 'output_endpoint'],
				'permission_callback' => __return_true(),
				'args'     => [
					'id'     => [
						'description' => 'Post ID.',
						'type'        => 'numeric',
					],
					'number' => [
						'description' => 'Number of posts',
						'type'        => 'numeric',
						'default'     => 5,
					],
				],
			]
		);
	}

	/**
	 * Output related posts endpoint
	 *
	 * @param  \WP_REST_Request $request REST request
	 * @since  3.2
	 * @return \WP_REST_Response
	 */
	public function output_endpoint($request)
	{
		$id = $request['id'];

		$posts          = $this->find_related($id, (int) $request['number']);
		$prepared_posts = [];

		if (!empty($posts)) {
			foreach ($posts as $post) {
				$prepared_post = [];

				$prepared_post['id']           = $post->ID;
				$prepared_post['link']         = get_permalink($post->ID);
				$prepared_post['status']       = $post->post_status;
				$prepared_post['title']        = [
					'raw'      => $post->post_title,
					'rendered' => get_the_title($post->ID),
				];
				$prepared_post['author']       = (int) $post->post_author;
				$prepared_post['parent']       = (int) $post->post_parent;
				$prepared_post['menu_order']   = (int) $post->menu_order;
				$prepared_post['content']      = [
					'rendered' => post_password_required($post) ? '' : apply_filters('the_content', $post->post_content),
				];
				$prepared_post['date']         = $post->post_date;
				$prepared_post['date_gmt']     = $post->post_date_gmt;
				$prepared_post['modified']     = $post->post_modified;
				$prepared_post['modified_gmt'] = $post->post_modified_gmt;

				$prepared_posts[] = $prepared_post;
			}
		}

		$response = new \WP_REST_Response();
		$response->set_data($prepared_posts);

		return $response;
	}

	/**
	 * Register gutenberg block
	 *
	 * @since  3.2
	 */
	public function register_block()
	{
		// Must be WP 5.0+
		if (!function_exists('register_block_type')) {
			return;
		}

		wp_register_script(
			'elasticpress-related-posts-block',
			EP_URL . 'dist/js/related-posts-block-script.min.js',
			[
				'wp-blocks',
				'wp-element',
				'wp-editor',
				'wp-api-fetch',
			],
			EP_VERSION,
			true
		);

		// The wp-edit-blocks style dependency is not needed on the front end of the site.
		$style_dependencies = is_admin() ? ['wp-edit-blocks'] : [];

		wp_register_style(
			'elasticpress-related-posts-block',
			EP_URL . 'dist/css/related-posts-block-styles.min.css',
			$style_dependencies,
			EP_VERSION
		);

		register_block_type(
			'elasticpress/related-posts',
			[
				'attributes'      => [
					'number' => [
						'type'    => 'number',
						'default' => 5,
					],
					'align'  => [
						'type' => 'string',
						'enum' => ['left', 'center', 'right', 'wide', 'full'],
					],
				],
				'editor_script'   => 'elasticpress-related-posts-block',
				'editor_style'    => 'elasticpress-related-posts-block',
				'style'           => 'elasticpress-related-posts-block',
				'render_callback' => [$this, 'render_block'],
			]
		);
	}

	/**
	 * Render Gutenberg block
	 *
	 * @param  array $attributes Block attributes
	 * @since  3.2
	 * @return string
	 */
	public function render_block($attributes)
	{
		$posts = $this->find_related(get_the_ID(), $attributes['number']);

		if (empty($posts)) {
			return '';
		}

		$class = 'wp-block-elasticpress-related-posts';

		if (!empty($attributes['align'])) {
			$class .= ' align' . $attributes['align'];
		}

		if (!empty($attributes['className'])) {
			$class .= ' ' . $attributes['className'];
		}

		ob_start();
?>
		<section class="<?php echo esc_attr($class); ?>">
			<ul>
				<?php foreach ($posts as $related_post) : ?>
					<li>
						<a href="<?php echo esc_url(get_permalink($related_post->ID)); ?>">
							<?php echo wp_kses(get_the_title($related_post->ID), 'ep-html'); ?>
						</a>
					</li>
				<?php endforeach; ?>
			</ul>
		</section>
	<?php

		$block_content = ob_get_clean();

		return $block_content;
	}

	/**
	 * Register related posts widget
	 *
	 * @since  2.2
	 */
	public function register_widget()
	{
		register_widget(__NAMESPACE__ . '\Widget');
	}

	/**
	 * Output feature box summary
	 *
	 * @since 2.1
	 */
	public function output_feature_box_summary()
	{
	?>
		<p><?php esc_html_e('ElasticPress understands data in real time, so it can instantly deliver engaging and precise related content with no impact on site performance.', 'elasticpress'); ?></p>
	<?php
	}

	/**
	 * Output feature box long
	 *
	 * @since 2.1
	 */
	public function output_feature_box_long()
	{
	?>
		<p><?php echo wp_kses_post(__('Output related content using our Widget or directly in your theme using our <a href="https://github.com/10up/ElasticPress/#related-posts">API functions.</a>', 'elasticpress')); ?></p>
<?php
	}
}
